finetrainers/ltx_video/ltx_video_lora.py

```diff
@@ -187,6 +348,32 @@ def collate_fn_t2v(batch: List[List[Dict[str, torch.Tensor]]]) -> Dict[str, torc
     }
 
 
+def collate_fn_i2v(batch: List[List[Dict[str, torch.Tensor]]]) -> Dict[str, torch.Tensor]:
+    return {
+        "prompts": [x["prompt"] for x in batch[0]],
+        "videos": torch.stack([x["video"] for x in batch[0]]),
+    }
+
+
+def prepare_noisy_latents_i2v(
+    sigmas: torch.Tensor,
+    noise: torch.Tensor,
+    latents: torch.Tensor,
+    num_frames: int,
+    height: int,
+    width: int,
+    **kwargs,
+) -> torch.Tensor:
+    # We do not add noise to the first frame as it is what we want to condition on.
+    image_frame_end_offset = 1 * height * width
+    image_latents = latents[:, :image_frame_end_offset]
+    video_noisy_latents = (1.0 - sigmas) * latents[:, image_frame_end_offset:] + sigmas * noise[
+        :, image_frame_end_offset:
+    ]
+    noisy_latents = torch.cat([image_latents, video_noisy_latents], dim=1)
```

Comment:
Shouldn't this be `noisy_latents = torch.cat([video_noisy_latents, image_latents], dim=1)`?

This is what we do in Cog, too:

https://github.com/a-r-r-o-w/finetrainers/blob/2a594ea2dbea5e9701c1c3494524a567f3071f83/training/cogvideox/cogvideox_image_to_video_lora.py#L806

Reply:
Wow, that is cool. Thanks for explaining. Then let’s make this a note?

Reply:
This and [this](https://github.com/a-r-r-o-w/finetrainers/pull/150/#discussion_r1897880866) -- suggests we should definitely experiment if noising the conditional latent is a good regularizer in this case.

Reply:
@yoavhacohen I made an assumption that the image latent is always kept clean and no denoising is applied to it. Is that the case when training LTXV? Or do you add a varying amount of noise based on some randomly sampled timestep (different from that of other frame latents) and perform denoising on it too? This is the first time I've seen this technique of per-token/per-frame denoising level so I'm not sure what to do without making guesses :/

Reply:
I recommend training text-to-video and image-to-video models simultaneously, adding a varying amount of noise based on a randomly sampled small timestep.
For training with image conditioning, you just need to determine the noise scheduler to apply to the tokens that correspond to the conditioning frame (versus the other tokens).
I recommend making the implementation generic rather than specific to the first frame - it should support conditioning on any subset of tokens, not just those corresponding to the first frame.

Reply:
Thank you! And generally speaking, the noise to be added to the conditioning frame — should it have a smaller magnitude than the one being added to the rest of the tokens?

Reply:
@yoavhacohen That sounds great, thank you! Since we don't support the idea of per-token timesteps in diffusers, I think we might have to write a custom scheduler step implementation - will give this a stab soon

Reply:
The noise applied to conditioning tokens should be reduced compared to other tokens in the sequence - that's actually what defines them as conditioning tokens. You can think of this as a kind of soft mask.​​​​​​​​​​​​​​​​